### PR TITLE
Replace webapp deployment script with faster and more flexible functions

### PR DESCRIPTION
**Pull request by lupusbytes**
#### This PR replaces azure-cli/webapp/deploy.ps1 with Initialize-WebApp.ps1.

`Initialize-WebApp` uses JMESPath wherever possible to check the states and avoid unnecessary work, which increases the performance more than tenfold while having complete feature parity.

Other notable changes:
- KeyVault secret permission configurement is now optional
- CORS/Allowed Origins is now optional and can be a wildcard or multiple domains instead of being a hardcoded to wildcard
- Optional VNet integration parameter
- Old/stale appsettings will be automitcally removed each run

### CODE REVIEW

**Review by kimlundjohansen — DISMISSED** [review]

File: azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1 [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    [Parameter(Mandatory = $true)]
+    [string]
+    $ObjectId,
+
+    [Parameter(Mandatory = $true)]
+    [ValidateSet("all", "backup", "delete", "get", "list", "purge", "recover", "restore", "set")]
+    [string[]]
+    $SecretPermissions,
+
+    [Parameter(Mandatory = $true)]
```
Comment (perkops) [inline azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1:17]: Would an alias make sense here?

File: azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1 [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    [string]
+    $ResourceGroupName
+  )
+
+  Write-Host "  Verifiying secret permissions '$SecretPermissions' on Key Vault '$KeyVaultName' for Object ID '$ObjectId'" -ForegroundColor DarkYellow -NoNewline
+
+  $output = az keyvault list `
+    --resource-group $ResourceGroupName `
+    --query "[?name=='$($KeyVaultName)']|[0].properties.accessPolicies[?objectId=='$ObjectId']|[0].permissions.secrets"
```
Comment (perkops) [inline azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1]: Dont actually think the double wrapping was needed?

File: azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1 [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  # Create a boolean that determines if we should invoke the `az keyvault set-policy` command.
+  # If the $output variable is null, it means that there is no pre-existing policies configured and thus we should always invoke the command.
```
Comment (perkops) [inline azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1]: are no*

File: azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1 [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    # If $invokeKeyVaultSetPolicy is false, we know that we got a match and a valid JSON response from the query above.
+    $existingPermissions = $output | ConvertFrom-Json
+
+    # Use LINQ to determine if all the permissions in our $SecretPermissions parameter is already present on the key vault.
+    $invokeSetKeyVaultPolicy = ![Linq.Enumerable]::All($SecretPermissions, [Func[string, bool]] { param($x) $existingPermissions.Contains($x) })
+  }
+
+  if ($invokeSetKeyVaultPolicy) {
+    Write-Host "  -> Permissions are not configured." -ForegroundColor Cyan
+    Write-Host "  Granting '$SecretPermissions' secret permissions to target Key Vault '$KeyVaultName'" -ForegroundColor DarkYellow
```
Comment (perkops) [inline azure-cli/keyvault/Set-KeyVaultSecretPermissions.ps1]: superfluous word "target"

File: azure-cli/utilities/VnetIntegration.ps1 [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class VnetIntegration {
+  [Parameter(Mandatory = $true)]
+  [string]
+  $Vnet
```
Comment (perkops) [inline azure-cli/utilities/VnetIntegration.ps1]: VNetName?

File: azure-cli/utilities/VnetIntegration.ps1 [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+  [Parameter(Mandatory = $true)]
+  [string]
+  $Subnet
```
Comment (perkops) [inline azure-cli/utilities/VnetIntegration.ps1]: SubnetName?

File: azure-cli/webapp/Get-WebAppManagedIdentityPrincipalId.ps1 [hunk 7]
```diff
@@ -0,0 +1,34 @@
+function Get-WebAppManagedIdentityPrincipalId {
+  param(
+    [Parameter(Mandatory = $true)]
+    [string]
+    $WebAppName,
```
Comment (perkops) [inline azure-cli/webapp/Get-WebAppManagedIdentityPrincipalId.ps1:6]: Alias here aswell?

File: azure-cli/webapp/Initialize-WebApp.ps1 [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    New-WebApp `
+      -Name $WebAppName `
+      -AppServicePlanId $AppServicePlanId `
+      -ResourceGroupName $ResourceGroupName `
+      -ResourceTags $ResourceTags
+
+    Throw-WhenError -output $output
+
+    # As with `az webapp list`, `az webapp create` returns null values for all siteConfig properties.
+    # Therefor we invoke the `az webapp show` query again, to get all the needed info.
```
Comment (perkops) [inline azure-cli/webapp/Initialize-WebApp.ps1]: Therefore
REF: https://www.grammarly.com/blog/therefore-vs-therefor/

File: azure-cli/webapp/Initialize-WebApp.ps1 [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+  #############################################################################################
+  # Ensure correct siteConfig settings
+  #############################################################################################
+  Write-Host "  Verifying platform and TLS configuration" -ForegroundColor DarkYellow -NoNewline
+
+  $webAppResource = $webAppResourceJson | ConvertFrom-Json -AsHashtable
+
+  if ($webAppResource.minTlsVersion -ne $minTlsVersion -or
+    $webAppResource.use32BitWorkerProcess -ne $false) {
+    Write-Host " -> Changes changes found." -ForegroundColor Cyan
```
Comment (perkops) [inline azure-cli/webapp/Initialize-WebApp.ps1]: Remove superfluos word "changes"

File: azure-cli/webapp/Initialize-WebApp.ps1 [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    Write-Host "  Updating general settings" -ForegroundColor DarkYellow
+    $output = az webapp config set `
+      --name $WebAppName `
+      --resource-group $ResourceGroupName `
+      --min-tls-version $MinTlsVersion `
+      --ftps-state Disabled `
```
Comment (perkops) [inline azure-cli/webapp/Initialize-WebApp.ps1:118]: Should we maybe include the ftps state in the comparison?

File: azure-cli/webapp/New-WebApp.ps1 [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    [Alias("Plan")]
+    [string]
+    $AppServicePlanId,
+
+    [Parameter(Mandatory = $true)]
+    [ValidateNotNullOrEmpty()]
+    [string]
+    $ResourceGroupName,
+
+    [Parameter(Mandatory = $false)]
```
Comment (perkops) [inline azure-cli/webapp/New-WebApp.ps1:20]: We should be able to generate a validationSet for this property:

az webapp list-runtimes
```

[
  "aspnet|V4.8",
  "aspnet|V3.5",
  "DOTNETCORE|2.1",
  "DOTNETCORE|3.1",
  "DOTNET|5.0",
  "DOTNET|6.0",
  "node|10.6",
  "node|10.10",
  "node|10.14",
  "node|12-lts",
  "node|14-lts",
  "php|7.4",
  "java|1.8|Tomcat|7.0",
  "java|1.8|Tomcat|8.5",
  "java|1.8|Tomcat|9.0",
  "java|1.8|Java SE|8",
  "java|7|Tomcat|7.0",
  "java|7|Tomcat|8.5",
  "java|7|Tomcat|9.0",
  "java|7|Java SE|8",
  "java|11|Tomcat|7.0",
  "java|11|Tomcat|8.5",
  "java|11|Tomcat|9.0",
  "java|11|Java SE|8"
]
```

File: azure-cli/webapp/Set-WebAppCors.ps1 [hunk 12]
```diff
@@ -0,0 +1,74 @@
+function Set-WebAppCors {
+  param (
+    [Parameter(Mandatory = $true)]
+    [ValidateNotNullOrEmpty()]
+    [Alias("Name", "AppName")]
+    [string]
+    $WebAppName,
+
+    [Parameter(Mandatory = $false)]
```
Comment (perkops) [inline azure-cli/webapp/Set-WebAppCors.ps1]: *Nit: Move down, since not required parameter?

File: azure-cli/webapp/Set-WebAppCors.ps1 [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  Throw-WhenError -output $output
+
+  $existingAllowedOrigins = $output | ConvertFrom-Json
+
+  if ($AllowedOrigins.Count -eq 0) {
+    if ($null -eq $existingAllowedOrigins) {
+      Write-Host " -> CORS is not set, as intended." -ForegroundColor Cyan
+    }
+    else {
+      Write-Host " -> CORS is set, but is not intended." -ForegroundColor Cyan
```
Comment (perkops) [inline azure-cli/webapp/Set-WebAppCors.ps1:34]: but is not set as intended

File: azure-cli/webapp/Set-WebAppCors.ps1 [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      Write-Host "  Removing Allowed Origins '$existingAllowedOrigins'" -ForegroundColor DarkYellow
```
Comment (perkops) [inline azure-cli/webapp/Set-WebAppCors.ps1:35]: Is this looking nice in output? Or should we string.join it together?

File: azure-cli/webapp/Set-WebAppCors.ps1 [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        --name $WebAppName `
+        --resource-group $ResourceGroupName `
+        --allowed-origins
+    }
+  }
+  else {
+    if ($null -eq $existingAllowedOrigins -or `
+      (Compare-Object $existingAllowedOrigins $AllowedOrigins)) {
+      Write-Host " -> CORS changes detected." -ForegroundColor Cyan
+      Write-Host "  Setting CORS Allowed Origins to '$AllowedOrigins'" -ForegroundColor DarkYellow
```
Comment (perkops) [inline azure-cli/webapp/Set-WebAppCors.ps1:46]: Same question here: string.join? Or output is fine as is?

File: azure-cli/webapp/Set-WebAppCors.ps1 [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+      # If allowed-origins is a wildcard, we have to wrap and escape the wildcard for it work with az cli
```
Comment (perkops) [inline azure-cli/webapp/Set-WebAppCors.ps1]: Maybe append something like:  (when running in a linux pipeline) ??

**Review by perkops — DISMISSED** [review]

**Review by perkops — APPROVED** [review]

**Pull request merged**